### PR TITLE
Introduce initial schemas in IBMQ Provider component

### PR DESCRIPTION
**Pull request by yeralin**
# Summary
Implements #289
Initial work to introduce schemas in IBMQ Provider component.

### Details and comments
This PR includes first two basic schemas `User` and `Token` that is used in `qiskit.provider.ibmq.api_v2.rest.auth` package.

### CODE REVIEW

File: qiskit/providers/ibmq/api_v2/rest/schemas/auth.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        self.token_id = token_id
+
+        super().__init__(**kwargs)
+
+
+class UserSchema(BaseSchema):
+    """Schema for User."""
+
+    # Required properties.
+    urls = Dict(Url(), required=True)
```
Comment (diego-plan9) [inline qiskit/providers/ibmq/api_v2/rest/schemas/auth.py]: Can you detail more this `urls` field by changing it into a Schema of its own that becomes `Nested`? This is one of the things we want to achieve via the documentation effort - currently, the provider expects the data that comes from this method contains at least two keys with specific names.

File: qiskit/providers/ibmq/api_v2/rest/schemas/auth.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+"""Model and schema for authentication."""
+from marshmallow.fields import String, Url, Dict
+from qiskit.validation import BaseSchema, bind_schema, BaseModel
+
+
+class TokenSchema(BaseSchema):
+    """Schema for Token."""
+
+    # Required properties.
+    token_id = String('id', required=True)
```
[the remark this answers is not in the document]
Comment (yeralin) [reply qiskit/providers/ibmq/api_v2/rest/schemas/auth.py]: On the second thought, I think the build will fail if I set it to `id`. I'll give it a try

File: qiskit/providers/ibmq/api_v2/rest/schemas/auth.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+@bind_schema(TokenSchema)
+class Token(BaseModel):
+    """Model for Token.
+
+    Please note that this class only describes the required fields. For the
+    full description of the model, please check ``TokenSchema``.
+
+    Attributes:
+        token_id (str): long term access token.
```
[the remark this answers is not in the document]
Comment (yeralin) [reply qiskit/providers/ibmq/api_v2/rest/schemas/auth.py]: There is no `description` attribute or anything similar, however I found [this](https://marshmallow.readthedocs.io/en/3.0/api_reference.html#marshmallow.fields.Field)
```
metadata – Extra arguments to be stored as metadata.
```

[the remark this answers is not in the document]
Comment (yeralin) [reply qiskit/providers/ibmq/api_v2/rest/schemas/auth.py]: The information was retrieved from the swagger page: 
https://auth-dev.quantum-computing.ibm.com/explorer/#!/UserBaseApiModel/UserBaseApiModel_loginWithApiToken

File: qiskit/providers/ibmq/api_v2/rest/schemas/auth.py (hunk 1 above)
Comment (yeralin) [reply qiskit/providers/ibmq/api_v2/rest/schemas/auth.py]: There is a problem with the application of `Nested`.

Initially, I was trying to express this schema this way:
```
class UserApiUrlResponseSchema(BaseSchema):
    """Nested schema for UserInfoResponse"""

    # Required properties.
    protocol = fields.String(required=True, metadata='api protocol')
    url = fields.Url(required=True, metadata='URL address to the api')


class UserInfoResponseSchema(BaseSchema):
    """Schema for UserInfoResponse."""

    # Required properties.
    urls = fields.Nested(UserApiUrlResponseSchema, many=True, required=True,
                         metadata='urls and corresponding protocols associated with APIs')
```

But then I realized that those URLs are stored as a map:
```
{
    ...,
    "urls": {
            "http": "https://api.quantum-computing.ibm.com/api", 
            "ws": "wss://wss.quantum-computing.ibm.com/"
    },
    ...
}
```

In order for my schema to be deserialized correctly, it should have been looked like this:
```
{
    ...,
    "urls": {
            [
                { "protocol": "http", "url": "https://api.quantum-computing.ibm.com/api" },
                { "protocol": "ws", "url": "wss://wss.quantum-computing.ibm.com/"}
            ]
    },
    ...
}
```

That's why nested schema approach might not serve for our purposes.

Thinking about possible solutions...

Comment (yeralin) [reply qiskit/providers/ibmq/api_v2/rest/schemas/auth.py]: Found similar issue raised under marshmallow project:
https://github.com/marshmallow-code/marshmallow/issues/483

Comment (yeralin) [reply qiskit/providers/ibmq/api_v2/rest/schemas/auth.py]: Ok, I think I found a right combination:
```
class UserApiUrlResponseSchema(BaseSchema):
    """Nested schema for UserInfoResponse"""

    # Required properties.
    protocol = fields.String(required=True, metadata='api protocol')
    url = fields.Url(required=True, metadata='URL address to the api')


class UserInfoResponseSchema(BaseSchema):
    """Schema for UserInfoResponse."""

    # Required properties.
    urls = fields.Dict(fields.Nested(UserApiUrlResponseSchema), many=True, required=True,
                       metadata='urls and corresponding protocols associated with APIs')
```

File: qiskit/providers/ibmq/api_v2/rest/schemas/auth.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (yeralin) [reply qiskit/providers/ibmq/api_v2/rest/schemas/auth.py]: Yep, as expected the build failed with:
```
************* Module ibmq.api_v2.rest.schemas.auth
qiskit/providers/ibmq/api_v2/rest/schemas/auth.py:32:4: C0103: Class attribute name "id" doesn't conform to '([A-Za-z_][A-Za-z0-9_]{2,30}|(__.*__))$' pattern (invalid-name)
qiskit/providers/ibmq/api_v2/rest/schemas/auth.py:74:8: C0103: Attribute name "id" doesn't conform to '[a-z_][a-z0-9_]{2,30}$' pattern (invalid-name)
qiskit/providers/ibmq/api_v2/rest/schemas/auth.py:73:23: W0622: Redefining built-in 'id' (redefined-builtin)
```

[the remark this answers is not in the document]
Comment (yeralin) [reply qiskit/providers/ibmq/api_v2/rest/schemas/auth.py]: I guess I could suppress the warning?

File: qiskit/providers/ibmq/api_v2/rest/schemas/auth.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+class TokenSchema(BaseSchema):
+    """Schema for Token."""
+
+    # Required properties.
+    token_id = String('id', required=True)
+
+
+@bind_schema(TokenSchema)
+class Token(BaseModel):
```
[the remark this answers is not in the document]
Comment (diego-plan9) [reply qiskit/providers/ibmq/api_v2/rest/schemas/auth.py]: Re-ping.

**Comment by diego-plan9** [comment]
Thanks  @yeralin ! I added some tweaks before merging, for making future schemas a bit more homogeneous:

* renamed `LoginTokenRequestSchema` to `LoginRequestSchema` - this way we keep the first bit of the name similar to the """canonical""" name of the endpoint used in the url_maps (`login` -> `LoginRequestSchema`; `user_info` -> `UserInfoRequestSchema`).
* reorganized the order of the schemas - as other schemas might have a number of nested schemas, let's keep the "parent" schemas at the end of the files, for readability.

So finally merging! One thing to revise would be the naming convention for the helper schemas - I did not change it in this PR, but when there are more schema that contain nested schemas it would be nice to give it another look and see if we can either come up with a pattern that makes the names informative but not hyper long.

**Pull request merged**